protocol/encoder/encoder.go

```diff
@@ -0,0 +1,261 @@
+package encoder
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/codecrafters-io/kafka-tester/protocol/errors"
+)
+
+type Encoder struct {
+	bytes  []byte
+	offset int
+}
+
+func (re *Encoder) Init(raw []byte) {
+	re.bytes = raw
+	re.offset = 0
+}
+
+func (re *Encoder) PutRawBytesAt(in []byte, offset int, length int) {
+	copy(re.bytes[offset:offset+length], in)
+}
+
+func (re *Encoder) PutInt32At(in int32, offset int, length int) {
+	binary.BigEndian.PutUint32(re.bytes[offset:offset+length], uint32(in))
+}
+
+func (re *Encoder) PutUVarintAt(in uint64, offset int) {
+	binary.PutUvarint(re.bytes[offset:], in)
+}
+
+func (re *Encoder) PutVarintAt(in int64, offset int) {
+	binary.PutVarint(re.bytes[offset:], in)
+}
+
+// primitives
+
+func (re *Encoder) PutInt8(in int8) {
```

Comment:
```suggestion
func (re *Encoder) WriteInt8(in int8) {
```

(Let's use `WriteXYZ` instead of `PutXYZ` - parallel to Decoder's `Read`)